### DISCUSSION TITLE
Profile attribute multiplicity

### DISCUSSION THREAD

**Discussion by scottrudy** [cimug-org/CIMTool#106] [General]
I know there is a rule that states attributes must have a multiplicity of `0..1` in the model. I noticed that the CImTool also seems to enforce that restriction on profiles. So in order to get a type in a Profile with a `<<Primitive>>` or `<<Compound>>` stereotype to show up as an XSD element where the `maxOccurs="unbounded"` I'm forced to create an association in the model. Is that by design?

**Comment by tviegut** [comment]
Hi @scottrudy:   Thanks for posting your question and its a good one that I'm sure others may be interested in the answer as well.  First off, the CIM itself, by design, is intended to serve purely as an information model for our particular domain.  As such, in this CIM "information world", we also have de facto rules and conventions that we enforce that all of our supporting tooling is expected to adhere to as well.  This is one of those areas.  So, to directly address your question the answer is "Yes", CIMTool does enforce at the profiling level this rule of only 0..1 or 1..1 (i.e. enforces a max cardinality of 1) for attributes within a class (see:  ).  Hopefully this helps.  Again, thanks for using the discussion forum to get answers to your questions.  It's helps the community in general so we appreciate it.

As for the use of associations to these types of classes, this is not allowed.  From the modeling guide:  

![image](https://github.com/cimug-org/CIMTool/assets/3385173/18371c0d-a101-4393-ae50-4c45116ca4a8)

**Reply by scottrudy** [reply]
Thanks @tviegut. I am currently trying to reverse a client XSD schema, that was based off the CIM, into a model so that I can forward engineer the schema again. One of the situations I have is the use of phone number in the XSD schema. An entity could have different kinds of phone numbers, so instead of explicitly listing each kind of phone each entity can hold, this schema extends the `TelephoneNumber` Compound class to have a `type` attribute which is a `String`. There are several other complex types in the XSD that use this extended type as a list.

```xml
<xs:complexType name="Person">
    <xs:sequence>
        <xs:element name="phoneList" type="TelephoneNumber" minOccurs="0" maxOccurs="unbounded"/>
    </xs:sequence>
</xs:complexType>
```

Given the multiplicity attribute rule and the association rule you mentioned above, I'm a little lost on how this can be represented in a model. Would it be appropriate to subclass a Compound type into a class with a new attribute and, given the example above, create an ExPerson superclass for Person that has an association to that new subclassed type? Something like this

```mermaid
classDiagram      
    ExPerson <|-- Person
    TelephoneNumber <|-- ExTelephoneNumber
    ExPerson "0..*" -- "0..*" ExTelephoneNumber
    class Person{
      +String firstName
      +...
    }
    class ExPerson{
    }
    class TelephoneNumber{
      +String localNumber
      +...
    }
    class ExTelephoneNumber{
      +String type
    }    
```

Or is there another preferred solution in this case?

**Reply by tviegut** [reply]
@scottrudy :  Do you know if the XSD schema profile you are viewing was generated using CIMTool?  Or, regardless of tooling, do you know if it is an XSD schema that is IEC/CIM compliant (i.e. confirm to the IEC 62361-100 standard for XSD schema definitions)?   If so you should expect some things similar to what is shown below in this sample XSD schema.  If so you would probably catch in your complex type a `sawsdl:` reference such as:  `<xs:complexType name="Person" sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#Person">` which would let us know what release of the CIM that they XSD schema was derived from.  That would be quite helpful for your purposes and speed up what you're trying to doing.

```XML
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema xmlns:m="http://www.ucaiug.org/CIM/Examples/2024/Personnel#"
           xmlns:sawsdl="http://www.w3.org/ns/sawsdl"
           xmlns:xs="http://www.w3.org/2001/XMLSchema"
           targetNamespace="http://www.ucaiug.org/CIM/Examples/2024/Personnel#"
           elementFormDefault="qualified"
           attributeFormDefault="unqualified">
   <xs:annotation>
      <xs:documentation xml:lang="en">
	Copyright 2024 UCAIug

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

	https://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	
	See the License for the specific language governing permissions and
	limitations under the License.
</xs:documentation>
   </xs:annotation>
   <xs:annotation/>
   <xs:complexType name="Person" sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#Person">
      <xs:annotation>
         <xs:documentation>General purpose information for name and other information to contact people.</xs:documentation>
      </xs:annotation>
      <xs:sequence>
         <xs:element name="firstName" minOccurs="0" maxOccurs="1"  type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#Person.firstName">
         </xs:element>
         <xs:element name="lastName" minOccurs="0" maxOccurs="1" type="xs:string"
                    sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#Person.lastName">
            <xs:annotation>
               <xs:documentation>Person's last (family, sir) name.</xs:documentation>
            </xs:annotation>
         </xs:element>
         <xs:element name="landlinePhone" minOccurs="0" maxOccurs="1" type="m:TelephoneNumber"
                     sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#Person.landlinePhone">
            <xs:annotation>
               <xs:documentation>Landline phone number.</xs:documentation>
            </xs:annotation>
         </xs:element>
      </xs:sequence>
   </xs:complexType>
   <xs:complexType name="TelephoneNumber" sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#TelephoneNumber">
      <xs:annotation>
         <xs:documentation>Telephone number.</xs:documentation>
      </xs:annotation>
      <xs:sequence>
         <xs:element name="areaCode" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#TelephoneNumber.areaCode">
            <xs:annotation>
               <xs:documentation>(if applicable) Area or region code.</xs:documentation>
            </xs:annotation>
         </xs:element>
         <xs:element name="cityCode" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#TelephoneNumber.cityCode">
            <xs:annotation>
               <xs:documentation>City code.</xs:documentation>
            </xs:annotation>
         </xs:element>
         <xs:element name="extension" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/2013/CIM-schema-cim16#TelephoneNumber.extension">
            <xs:annotation>
               <xs:documentation>(if applicable) Extension for this telephone number.</xs:documentation>
            </xs:annotation>
         </xs:element>
         . . . 
         . . . 
   </xs:complexType>
</xs:schema>
```

**Reply by tviegut** [reply]
@scottrudy :  I just dropped you an email at your yahoo account.  If that fits your availability I can set something up and we can include the outcomes in this thread.

**Reply by scottrudy** [reply]
Unfortunately, wherever it came from seems to be lost, and there is no model reference attribute. I am currently trying to map back to the latest model I could find, which uses iec61970cim17v40, iec61968cim13v13b, and iec62325cim03v17b. Everything I see in the forward generating XSD from that is showing `http://iec.ch/TC57/CIM-generic#{classname}` at the moment.

**Reply by scottrudy** [reply]
@tviegut : I dove back into this today and after further evaluation, the solution we discussed didn't give me exactly what I need to preserve backward compatibility because it introduces a wrapper class. I ended up doing what you see below and am just wondering if there are any issues with subclassing a Compound class.

```mermaid
classDiagram      
    ExPerson <|-- Person : CIMExtension
    ExTelephoneNumber <|-- TelephoneNumber  : CIMExtension
    TelephoneNumber <|-- TelephoneEntry
    ExPerson "1..*" -- "0..*" TelephoneEntry
    namespace Common {
        class Person{
            +String firstName
            +...
        }
        class TelephoneNumber{
        <<Compound>>
            +String localNumber
            +...
        }
    }
    class ExPerson{
        +TelephoneEntry phone
    }
    class TelephoneEntry{
    }
    class ExTelephoneNumber{
    <<Compound>>
        +String type
    }
```
With a few tweaks in the profile for naming, that then generates and XSD that looks similar to this (I still need to work on the namespacing):

```xml
<xs:schema>
   <xs:complexType name="Person" sawsdl:modelReference="http://iec.ch/TC57/CIM100#Person">
      <xs:sequence>
         <xs:element name="phone" minOccurs="0" maxOccurs="unbounded" type="m:TelephoneNumber"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#ExPerson.TelephoneEntry">
         </xs:element>
      </xs:sequence>
   </xs:complexType>
   <xs:complexType name="TelephoneNumber" sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneEntry">
      <xs:sequence>
         <xs:element name="areaCode" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.areaCode">
         </xs:element>
         <xs:element name="cityCode" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.cityCode">
         </xs:element>
         <xs:element name="countryCode" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.countryCode">
         </xs:element>
         <xs:element name="dialOut" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.dialOut">
         </xs:element>
         <xs:element name="extension" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.extension">
         </xs:element>
         <xs:element name="internationalPrefix" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.internationalPrefix">
         </xs:element>
         <xs:element name="ituPhone" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.ituPhone">
         </xs:element>
         <xs:element name="localNumber" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#TelephoneNumber.localNumber">
         </xs:element>
         <xs:element name="type" minOccurs="0" maxOccurs="1" type="xs:string"
                     sawsdl:modelReference="http://iec.ch/TC57/CIM100#ExTelephoneNumber.type">
         </xs:element>
      </xs:sequence>
   </xs:complexType>
</xs:schema>
```

**Reply by tviegut** [reply]
@scottrudy :  Did you resolve this profile definition yet?  I have another approach you can utilize if not.  Let me know.

**Reply by scottrudy** [reply]
What I mentioned above seems to work, but I'm open to alternatives.

**Reply by scottrudy** [reply]
Upon further reading, I have found a document that says stereotyped classes may not participate in relationships, which of course includes inheritance. I think this may leave me no choice except to create a class that is not stereotyped, and looks just like the stereotyped class. Again, this is just for backward compatibility purposes. I realize there are better ways to do this moving forward, such as creating a wrapper class around the stereotyped class, which would allow the collection to be held.

**Reply by tviegut** [reply]
@scottrudy : what document are you referring to?

**Reply by scottrudy** [reply]
@tviegut [Application of IEC CIM Standards in Power System Modeling, Smart Grid and Enterprise integration](https://cimug.ucaiug.org/CIMGroups/CIMInterop/IEEE%20Tutorial/CIM%20Tutorial%20Master%20Slide%20Deck-2013.pdf), page 151 states:

> Stereotyped classes are types that never participate in relationships (i.e., no associations, no inheritance), but are used as types for attributes

**Reply by tviegut** [reply]
@scottrudy :  Thanks for the link.  I got a chance to check it out today.  I'm familiar with the presentation as it was at an IEEE conference probably 12 or 13 years ago.  I apologize as there are different contexts for different presentations and sometime it may not match the scenario you're attempting to apply.  What is described in that presentation was a formal approach for how to prepare a use case and corresponding UML modelling proposal in the form of extensions with the goal of having such extensions reviewed and formally added (normatively) to base CIM.  What is not explicitly called out or explained (but is implied) is that such a proposal would serve as interim UML model representation for extensions that help make a CIM Model Managers life much easier (i.e. I am a CIM Model Managers for Transmission) when it comes time to integrate/synthesize new attributes, classes, associations into the core CIM.  If I were to receive a separate package of extensions as described in the presentation and those extensions reviewed and broadly approved by the team I could easily migrated the relevant attributes, classes, associations and generalizations out of the extensions package and into the normative CIM model.  That's a key reason for the proposed format for those extensions.  They would not remain in the format you see in the presentation but rather the "shadow" classes would be deleted after relevant attributes and associations were either moved or "reassigned" to appropriate classes within the normative CIM.  In that context I wouldn't retain any stereotypes on these extensions upon migrating them into the CIM.  Anyway, this was my reason for asking for the source of what you were referencing so as to review the context.  Note that in CIMTool, an internal "union" of extensions into a base CIM model, in principal, achieves something roughly similar to what I'm describing as a manual merging. In any event, just following up on your response.
